scarb/src/compiler/plugin/proc_macro/host.rs

```diff
@@ -1172,3 +1173,66 @@ impl ProcMacroHost {
         &self.macros
     }
 }
+
+/// A global storage for dynamically-loaded procedural macros.
+/// Loads dynamic shared libraries and hides them beside [`ProcMacroHostPlugin`].
+/// Guarantees that every library is loaded exactly once,
+/// but does not prevent loading multiple versions of the same library.
+#[derive(Default)]
+pub struct ProcMacroRepository {
+    /// A mapping between the [`PackageId`] of the package which defines the plugin
+    /// and the [`ProcMacroHostPlugin`] holding the underlying shared library.
+    macros: HashMap<PackageId, Arc<ProcMacroHostPlugin>>,
+}
+
+impl ProcMacroRepository {
+    /// Inserts the pre-loaded [`ProcMacroInstance`] to the storage, without checking
+    /// whether some plugin for the [`PackageId`] already exists.
+    pub fn insert_prebuilt_unchecked(
+        &mut self,
+        package_id: PackageId,
+        instance: Arc<ProcMacroInstance>,
+    ) -> Result<()> {
+        let plugin = Arc::new(ProcMacroHostPlugin::try_new(vec![instance])?);
```

Comment:
We will need some validations on proc macro expansion duplicates, right now it's handled in the ProcMacroHostPlugin with assumption that all proc macros are added to a single plugin. See https://github.com/software-mansion/scarb/blob/b3558737984e987c746fe1794a08ce0dd3ac9172/scarb/src/compiler/plugin/proc_macro/host.rs#L138